### PR TITLE
[RESTEASY-1916]:Fix InboundSseEventImpl doesn't getReader from ClientConfiguration

### PR DESCRIPTION
**Pull request by jimma**
[RESTEASY-2011]:Allow disable JsonBindingProvider by setting resteasy.preferJacksonOverJsonB in client

### CODE REVIEW

**Comment by rsvoboda** [comment]
@marekkopecky please check this PR

**Comment by marekkopecky** [comment]
@rsvoboda I prefer to backport fix for RESTEASY-2011 to 3.6 branch

**Comment by marekkopecky** [comment]
This works, by in my PoV, this is user-unfriendly:
```
ResteasyClientBuilder builder = new ResteasyClientBuilder();
Configuration config = builder.property("resteasy.preferJacksonOverJsonB", true).getConfiguration();
ResteasyProviderFactory.pushContext(Configuration.class, config);
ResteasyClient client = builder.withConfig(config).build()
```

**Comment by jimma** [comment]
> https://issues.jboss.org/browse/RESTEASY-1916
Can this be solved by solution of RESTEASY-1939?

I think it doesn't. Client doesn't push any context data. 

>https://issues.jboss.org/browse/RESTEASY-2011
I check to use -Dproperty (java -Djava.util.logging.manager=org.jboss.logmanager.LogManager -Dlogging.configuration=file:logging.properties -Dresteasy.preferJacksonOverJsonB=true -jar target/RESTEasyClientApplication.jar) and client.property("resteasy.preferJacksonOverJsonB", true);. Both doesn't work and JsonBindingProvider is still used on client side (according to the logs). Can you please double check?

I doesn't add support for disable JsonBindingProvider by system property.  I'll add this one.  @asoldano  @rsearls  What do you think adding this system property support ? 

> First commit in this PR changes pom files, second commit revert pom changes, I prefer to do some cleanup of commits.
This will be addressed when this merged.

**Comment by jbliznak** [comment]
@jimma and what about this? Let the builder to handle the property?
``ResteasyClient client = new ResteasyClientBuilder().property("resteasy.preferJacksonOverJsonB", true).build();``

**Comment by marekkopecky** [comment]
> I doesn't add support for disable JsonBindingProvider by system property. I'll add this one. @asoldano @rsearls What do you think adding this system property support ?

Server supports system property for disabling JsonBindingProvider, we should be consistent.

File: testsuite/integration-tests/pom.xml [hunk 1]
```diff
@@ -504,11 +503,33 @@
...
+                            <configuration>
+                                <trimStackTrace>false</trimStackTrace>
+                                <skip>false</skip>
+                                <systemPropertyVariables combine.children="append">
+                                    <allow.gzip.server.jvm.arg>-Dresteasy.allowGzip=true</allow.gzip.server.jvm.arg>
+                                </systemPropertyVariables>
+                                <includes>
+                                    <include>**/AllowGzipOnServerAllowGzipOnClientTest.java</include>
+                                    <include>**/AllowGzipOnServerNotAllowGzipOnClientTest.java</include>
+                                </includes>
```
Comment (asoldano) [inline testsuite/integration-tests/pom.xml]: How are the gzip changes here related to the PR?

**Comment by marekkopecky** [comment]
Btw, this configuration currently (3.6.1.Final) works on client side:
```java
        CustomResteasyConfiguration config = new CustomResteasyConfiguration();
        ResteasyProviderFactory.pushContext(ResteasyConfiguration.class, config);
        ResteasyClient client = new ResteasyClientBuilder().build();
```
And CustomResteasyConfiguration:
```java
public class CustomResteasyConfiguration implements ResteasyConfiguration {

    public static Map<String, String> param = new HashMap<>();

    static {
        param.put("resteasy.preferJacksonOverJsonB", "true");
    }

    @Override
    public String getParameter(String name) {
        return param.get(name);
    }
...
```

**Comment by jimma** [comment]
@I'd rather have these 2 commits in different PRs, unless there's a strong reason for doing like this. Can you please update this PR to only fix 1916 and create another one for 2011 ?

The test for 1916 is rely on 2011.  It is testing against the customized jackson provider.

File: resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/providers/sse/InboundSseEventImpl.java [hunk 2]
```diff
@@ -188,13 +191,20 @@ public String readData()
    }
 
    public <T> T readData(GenericType<T> type, MediaType mediaType)
-   {
+   {  
       //System.out.println("Thread " + Thread.currentThread().getName() + "read data");
       final MediaType effectiveMediaType = mediaType == null ? this.mediaType : mediaType;
-      ResteasyProviderFactory factory = ResteasyProviderFactory.getInstance();
-      RegisterBuiltin.register(factory);
-      final MessageBodyReader reader = factory.getClientMessageBodyReader(type.getRawType(), type.getType(),
```
[the remark this answers is not in the document]
Comment (jimma) [reply resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/providers/sse/InboundSseEventImpl.java]: For client, there is no LocalResteasyProviderFactory we can retrieve with RestesyProviderFactory.getInstance(): 
In RestesyClientBuilder , LocalResteasyProviderFactory is constructed with : 
```
   protected ResteasyProviderFactory getProviderFactory()
   {
      if (providerFactory == null)
      {
         // create a new one
         providerFactory = new LocalResteasyProviderFactory(ResteasyProviderFactory.newInstance());
         RegisterBuiltin.register(providerFactory);
      }
      return providerFactory;
   }
```

**Comment by jimma** [comment]
>@jimma and what about this? Let the builder to handle the property?
ResteasyClient client = new ResteasyClientBuilder().property("resteasy.preferJacksonOverJsonB", true).build();
@jbliznak  This actually sets the property to LocalResteasyProviderFactory and there is no way to get this value from JsonBindingProvider. It now requires manually push a thread local configuration contains this value to ResteayProviderFactory :  ResteasyProviderFactory.pushContext(ResteasyConfiguration.class, config);

**Comment by jimma** [comment]
> Server supports system property for disabling JsonBindingProvider, we should be consistent.
From JsonBindingProvider, we don't have this support : 
https://github.com/resteasy/Resteasy/blob/master/providers/json-binding/src/main/java/org/jboss/resteasy/plugins/providers/jsonb/JsonBindingProvider.java#L50

File: resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/providers/sse/InboundSseEventImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (asoldano) [reply resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/providers/sse/InboundSseEventImpl.java]: Right. So my point is that maybe the factory should be retrieved differently there, maybe refactoring the way the InboundSseEventImpl is constructed, etc.

**Comment by asoldano** [comment]
@jimma 

> > I'd rather have these 2 commits in different PRs, unless there's a strong reason for doing like this. Can you please update this PR to only fix 1916 and create another one for 2011 ?
> 
> The test for 1916 is rely on 2011. It is testing against the customized jackson provider.

I noticed that, but I believe we should have an independent test for 1916.

File: resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/providers/sse/InboundSseEventImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimma) [reply resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/providers/sse/InboundSseEventImpl.java]: @asoldano  I thought of this, and SseEventSourceImpl can't get  LocalResteasyproviderFacotry from WebTarget.

**Comment by jimma** [comment]
> I noticed that, but I believe we should have an independent test for 1916.

@asoldano  Do you mean we add another exclusion in pom.xml to disable jsonbinding provider for RESTEASY-1916 test ?

**Comment by asoldano** [comment]
> @asoldano Do you mean we add another exclusion in pom.xml to disable jsonbinding provider for RESTEASY-1916 test ?

No, I simply mean that RESTEASY-1916 deserves a test that verifies it only; it's a bug reported in May, way before we created RESTEASY-2011, so please think about a way to test it alone (1916, I mean).

**Comment by jimma** [comment]
This will be replaced by https://github.com/resteasy/Resteasy/pull/1694

**Pull request closed without merging**